notes/smart-pointers-2.md

```diff
@@ -0,0 +1,210 @@
+# Smart Pointers, Part 2
+
+Contributors: Edward Zhang
+
+---
+## `std::shared_ptr`
+
+Let's look at shared pointers!
+
+These are similar to unique pointers, but now we allow objects to be shared / "owned" by multiple shared pointers
+
+---
+## Reference counting
+
+If multiple shared pointers can own an object, how do we know when to `delete` the pointed-to object?
+
+Quite intuitively, we use the notion of **reference counting**. For each object, we keep a count of how many shared pointers are pointing to it
+> NOTE: this section provides some nice intuition for reference counts. However, keep reading to see how complications can arise because of the actual implementation
+
+The reference count for an object is adjusted like so:
+- When we make a new  shared pointer to the object, increment its reference count by 1
+- When a shared pointer to it is destroyed, decrement reference count by 1
+
+Note that shared pointers can be copied
+- I.e., its copy constructor and copy-assignment operators are defined
+- Copy-assignment can also increment or decrement reference counts (see next section)
+
+Finally, when a shared pointer sees that the object's reference count is 0, destroy the pointed-to object
+
+---
```

Comment:
Perhaps its a good idea to show implementation first, as the following examples will make more sense.